Implement tag mapping in graph refresh

cben opened the pull request:
- [x] Depends on https://github.com/ManageIQ/manageiq/pull/16734 — merged

Subsumes #384, done with lots of help from Ladislav & Julian :earth_africa: 

https://bugzilla.redhat.com/show_bug.cgi?id=1506404

cc @bronaghs @djberg96 @Ladas @agrare (in case anybody still here, and happy holidays)

cben commented:
@juliancheal @Ladas This is still WIP but I think something like this...

Depends on couple core changes, PR in a minute.

cben commented:
rebased

cben commented on app/models/manageiq/providers/amazon/inventory_collection_default/cloud_manager.rb:
```diff
@@ -176,6 +176,30 @@ def vm_and_template_labels(extra_attributes = {})
+      }
+
+      attributes[:targeted_arel] = lambda do |inventory_collection|
+        ids = inventory_collection.parent_inventory_collections.collect_concat(&:id)
```
I was trying to "simplify" this to something like I did containers and avoid going through Vm.vm_and_template_taggings relation, just because it scared me.

But `add_inventory_collection` requires `:association => :foos` to be set, this determines the name of the `persister.foos` accessor to the collection :-(
Anyway, now that I added that relation in https://github.com/ManageIQ/manageiq/pull/16734 I should probably do what vm_and_template_labels does...

Ladas replied:
ids are not available here yet, so we need to use
```manager_uuids = inventory_collection.parent_inventory_collections.collect(&:manager_uuids).map(&:to_a).flatten```

manager uuids are :ems_ref columns of Vm in this case. So we need a nested query that transforms that to ids

Ladas commented on app/models/manageiq/providers/amazon/inventory_collection_default/cloud_manager.rb:
```diff
@@ -176,6 +176,30 @@ def vm_and_template_labels(extra_attributes = {})
+        ids = inventory_collection.parent_inventory_collections.collect_concat(&:id)
+        Tagging.where(
+          'taggable_type' => 'VmOrTemplate',
+          'taggable_id'   => ids
```
so then here it would be 
```
'taggable_id'  => inventory_collection.parent.vm_and_templates.where(:ems_ref => manager_uuids)
```

Ladas commented on app/models/manageiq/providers/amazon/inventory_collection_default/cloud_manager.rb:
```diff
@@ -190,11 +190,13 @@ def vm_and_template_taggings(extra_attributes = {})
           'taggable_type' => 'VmOrTemplate',
-          'taggable_id'   => ids
-        ).joins(:tag).merge(Tag.controlled_by_mapping)
+          'taggable_id'   => ems.vm_and_template_taggings.where(:ems_ref => manager_uuids)
```
just 
```
ems.vm_and_templates.where(:ems_ref => manager_uuids)
```

since the manager uuis are fro ma table VmOrTemplate

cben replied:
thanks, fixed

Ladas commented on spec/models/manageiq/providers/amazon/cloud_manager/vcr_specs/refresher_inventory_object_spec.rb:
```diff
@@ -32,6 +33,7 @@
           end
 
           assert_common
+          assert_mapped_tags_on_vm
```
could we move this inside of the specific Vms asserts?

Especially  assert_specific_vm_powered_on, assert_specific_vm_on_cloud_network_public_ip and assert_specific_vm_on_cloud_network

That way it will be automatically tested in old, new and targeted refresh

cben replied:
done for vms

For templates the only label I found at all is `Name` :neutral_face: 
and both `assert_specific_template*` have 0 labels (not even `Name`) so only calling this function explicitly in some tests.
This means targeted image taggings refresh is not well tested.

Ladas commented on spec/models/manageiq/providers/amazon/aws_refresher_spec_common.rb:
```diff
@@ -1416,9 +1416,27 @@ def assert_relationship_tree
-  #   expect(vm.tags).to be_empty
-  # end
+  def assert_specific_labels_on_vm
+    vm = ManageIQ::Providers::Amazon::CloudManager::Vm.find_by(:name => "ladas_test_5")
```
We are usually testing only Vms having EmsRefreshSpec prefix, those are the ones nobody is touching (so we can re-record VCR easily)

Ladas commented on app/models/manageiq/providers/amazon/inventory/parser/cloud_manager.rb:
```diff
@@ -4,10 +4,16 @@ class ManageIQ::Providers::Amazon::Inventory::Parser::CloudManager < ManageIQ::P
+    @tag_mapper = ContainerLabelTagMapping.mapper
+
     # The order of the below methods doesn't matter since they refer to each other using only lazy links
+    persister.collections[:tags_to_resolve] = @tag_mapper.tags_to_resolve_collection
```
cool, you already use custom IC here, I was about to ask about it :-)

Ladas replied:
@cben could we move the IC definition into the defaults, and the init into the Persitors?

cben replied:
I could.  Note that constructing the Mapper actually reads whole mappings table.
- Is it OK to do it each Persister initialization?
- Will it still be re-read each refresh? (eg. if it's only read once when refresh worker starts that would be bad, changing mapping would have no effect).  
- Conversely, can re-reading it each targetted refresh be a performance problem?
- Will Persister initialization remain a good place with collector/persister split?

Ladas replied:
hum, ok, if it reads the whole mapping table, I will need to look on it more closely tomorrow. We will need to limit it only to Vms and Templates that are part of the targeted refresh

Ladas replied:
- Is it OK to do it each Persister initialization?

Right now that would be the same as having it in parser, that is being called each refresh

- Will it still be re-read each refresh? (eg. if it's only read once when refresh worker starts that would be bad, changing mapping would have no effect).

Yes, right now, the persister defines ICs dynamically (that might change in the future, as some providers use static definition, but there can be always a dynamic part)

- Conversely, can re-reading it each targetted refresh be a performance problem?

Yes looking at this, it will definitely hit targeted refresh performance, with huge amount of tags present

- Will Persister initialization remain a good place with collector/persister split?

There will be some refactoring of the format, but the Persistor should remain mostly the same.

cben commented:
Hooray!  I was stuck on mysterious hardware,disk,network count mismatches, turned out I dropped a line in merging conflicts.
This works now for VMs :tada: 
Adding images.

cben commented:
PTAL.  Added images, addressed most comments, except https://github.com/ManageIQ/manageiq-providers-amazon/pull/382#discussion_r159463356

Don't understand how refresher_other_region_spec.rb passes tags check in assert_specific_vm when I didn't `create_tag_mapping`?

Ladas commented:
@cben yes, another copy of assert_specific_vm_powered_on. I think the specs are great now, we can ignore the other region specs

Ladas commented on app/models/manageiq/providers/amazon/inventory/parser/cloud_manager.rb:
```diff
@@ -4,10 +4,15 @@ class ManageIQ::Providers::Amazon::Inventory::Parser::CloudManager < ManageIQ::P
+    @tag_mapper = ContainerLabelTagMapping.mapper
+
     # The order of the below methods doesn't matter since they refer to each other using only lazy links
+    persister.collections[:tags_to_resolve] = @tag_mapper.tags_to_resolve_collection
```
@cben ok, so continuing from https://github.com/ManageIQ/manageiq-providers-amazon/pull/382#discussion_r159463356

lets move this to persister init, we can put shared method to ManageIQ::Providers::Amazon::Inventory::Persister and just call it in Cloud and TargetCollection Persisters for now

And we should there add a TODO for perf, that this will always load all ContainerLabelTagMapping (or other?). But this table is fairly small? It's the Tagging that is huge, right?

cben replied:
Mapper init is reading ContainerLabelTagMapping.all.  It's fairly small, 1 per mapping defined in UI, independent of inventory size and number of tags created/assigned.
Moving...

cben replied:
was about to ask about TargetCollection then saw you already anticipated that :-)
Done, PTAL.

cben commented:
I believe the "targeting" of the targeted_arel is not really tested.  It does create the expected taggings, but are we sure it doesn't delete other taggings?
However, I'll leave this to you folks to test better if you wish.

Targeted query, as diff from normal `ems.vm_and_template_taggings` query:
```diff
 SELECT "taggings".* FROM "taggings"
 INNER JOIN "tags" ON "tags"."id" = "taggings"."tag_id"
 LEFT OUTER JOIN "classifications" ON "classifications"."tag_id" = "tags"."id" 
 INNER JOIN "vms" ON "taggings"."taggable_id" = "vms"."id" 
 WHERE "vms"."ems_id" = 12000000009214 
   AND "taggings"."taggable_type" = 'VmOrTemplate' 
   AND (("tags"."name" LIKE '/managed/amazon:%') OR ("tags"."name" LIKE '/managed/kubernetes:%'))
   AND "classifications"."read_only" = 't' 
   AND ("classifications"."parent_id" != 0) 
+  AND "taggings"."taggable_type" = 'VmOrTemplate' 
+  AND "taggings"."taggable_id" IN (SELECT "vms"."id" FROM "vms" WHERE "vms"."ems_id" = 12000000009214)
```

cben commented:
Hmm, that last SELECT seems wrong!  I don't see a `ems_ref` constraint on vms.  Debugging...

The `taggable_type = 'VmOrTemplate'` constraint is duplicate, I'm going to I drop it from targeted_arel.

cben commented:
OK, this was silly.  Turns out `record.association(args)` (note missing `.where`, see last commit) ignores the args, sort-of.
<details>

The reason associations accept args at all is side effect of a deprecated interface, good thing it's going away:
```
[4] pry(main)> ems.hosts
  Host Load (3.0ms)  SELECT "hosts".* FROM "hosts" WHERE "hosts"."ems_id" = $1  [["ems_id", 7]]

[5] pry(main)> ems.hosts(:ems_ref => 'ignored')
DEPRECATION WARNING: Passing an argument to force an association to reload is now deprecated and will be removed in Rails 5.1. Please call `reload` on the result collection proxy instead. (called from <main> at (pry):5)
  Host Load (0.4ms)  SELECT "hosts".* FROM "hosts" WHERE "hosts"."ems_id" = $1  [["ems_id", 7]]
```

</details>

----

Fixed targeted query, again as diff from `ems.vm_and_template_taggings` query:
```diff
 SELECT "taggings".* FROM "taggings" 
 INNER JOIN "tags" ON "tags"."id" = "taggings"."tag_id" 
 LEFT OUTER JOIN "classifications" ON "classifications"."tag_id" = "tags"."id" 
 INNER JOIN "vms" ON "taggings"."taggable_id" = "vms"."id" 
 WHERE "vms"."ems_id" = 12000000010061 
   AND "taggings"."taggable_type" = 'VmOrTemplate' 
   AND (("tags"."name" LIKE '/managed/amazon:%') OR ("tags"."name" LIKE '/managed/kubernetes:%'))
   AND "classifications"."read_only" = 't' 
   AND ("classifications"."parent_id" != 0)
+  AND "taggings"."taggable_id" IN (SELECT "vms"."id" FROM "vms" 
+                                   WHERE "vms"."ems_id" = 12000000010061 
+                                     AND "vms"."ems_ref" IN ('i-c72af2f6', 'ami-2051294a'))
```

I think this supports the need for a targeted-after-full-doesn't-delete spec :-), but will leave that to you.
I believe this is ready now.

Ladas commented on app/models/manageiq/providers/amazon/inventory_collection_default/cloud_manager.rb:
```diff
@@ -193,7 +193,7 @@ def vm_and_template_taggings(extra_attributes = {})
         ems = inventory_collection.parent
         ems.vm_and_template_taggings.where(
-          'taggable_id'   => ems.vms_and_templates(:ems_ref => manager_uuids)
+          'taggable_id'   => ems.vms_and_templates.where(:ems_ref => manager_uuids)
```
ah :-o

agrare replied:
Ha that'll do it

juliancheal commented:
👍 thanks @cben

simaishi commented:
Gaprindashvili backport details:

```
$ git log -1
commit 7bdb89b65ad4f39dfe65ab60ef62d99ea2f820cc
Author: Ladislav Smola <lsmola@redhat.com>
Date:   Thu Jan 4 14:31:04 2018 +0100

    Merge pull request #382 from cben/tag-mapping
    
    Implement tag mapping in graph refresh
    (cherry picked from commit 53cc7b01248aa5addc0e8d36d1f1dc152de75acf)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1531291
```
